File: internal/hosts/hosts.go
package hosts

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/4ster-light/phocus/internal/dns"
)

// getHostsFilePath returns the appropriate hosts file path for the current OS
func getHostsFilePath() string {
	if runtime.GOOS == "windows" {
		return `C:\Windows\System32\drivers\etc\hosts`
	}
	return "/etc/hosts"
}

func BlockDomain(domain string) error {
	f, err := os.OpenFile(getHostsFilePath(), os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return fmt.Errorf("error opening hosts file: %v", err)
	}
	defer f.Close()

	if _, err := f.WriteString(fmt.Sprintf("127.0.0.1 %s\n", domain)); err != nil {
		return fmt.Errorf("error writing to hosts file: %v", err)
	}

	// Add a small delay before flushing DNS
	time.Sleep(100 * time.Millisecond)

	return dns.FlushDNS()
}

func UnblockDomains(domains []string) error {
	input, err := os.ReadFile(getHostsFilePath())
	if err != nil {
		return fmt.Errorf("error reading hosts file: %v", err)
	}

	lines := strings.Split(string(input), "\n")
	var newLines []string

	for _, line := range lines {
		if !containsBlockedDomain(line, domains) {
			newLines = append(newLines, line)
		}
	}

	output := strings.Join(newLines, "\n")
	if err := os.WriteFile(getHostsFilePath(), []byte(output), 0644); err != nil {
		return fmt.Errorf("error writing hosts file: %v", err)
	}

	// Add a small delay before flushing DNS
	time.Sleep(100 * time.Millisecond)

	return dns.FlushDNS()
}

func containsBlockedDomain(line string, domains []string) bool {
	for _, domain := range domains {
		if strings.Contains(line, domain) {
			return true
		}
	}
	return false
}

File: src/hosts.rs
use anyhow::{Context, Result};
use std::fs;
use std::path::PathBuf;
use std::time::Duration;
use tokio::time::sleep;
use crate::dns::flush_dns;
use std::io::Write;

fn get_hosts_path() -> PathBuf {
    if cfg!(windows) {
        PathBuf::from(r"C:\Windows\System32\drivers\etc\hosts")
    } else {
        PathBuf::from("/etc/hosts")
    }
}

pub async fn block_domain(domain: &str) -> Result<()> {
    let hosts_path = get_hosts_path();
    let entry = format!("127.0.0.1 {}\n", domain);

    fs::OpenOptions::new()
        .append(true)
        .open(&hosts_path)
        .context("Failed to open hosts file")?
        .write_all(entry.as_bytes())
        .context("Failed to write to hosts file")?;

    // Add a small delay before flushing DNS
    sleep(Duration::from_millis(100)).await;
    flush_dns().await
}

pub async fn unblock_domains(domains: &[String]) -> Result<()> {
    let hosts_path = get_hosts_path();
    let content = fs::read_to_string(&hosts_path).context("Failed to read hosts file")?;

    let new_content: String = content
        .lines()
        .filter(|line| !domains.iter().any(|domain| line.contains(domain)))
        .collect::<Vec<_>>()
        .join("\n");

    fs::write(&hosts_path, new_content).context("Failed to write hosts file")?;

    // Add a small delay before flushing DNS
    sleep(Duration::from_millis(100)).await;
    flush_dns().await
}
